corgidrp/data.py

```diff
@@ -551,10 +556,84 @@ def __init__(self, data_or_filepath, pri_hdr=None, ext_hdr=None, input_dataset=N
         if 'DATATYPE' not in self.ext_hdr or self.ext_hdr['DATATYPE'] != 'NonLinearityCalibration':
             raise ValueError("File that was loaded was not a NonLinearityCalibration file.")
 
+class KGain(Image):
+    """
+    Class for KGain calibration file. Until further insights it is just one float value.
+
+    Args:
+        data_or_filepath (str or np.array): either the filepath to the FITS file to read in OR the calibration data. See above for the required format.
+        pri_hdr (astropy.io.fits.Header): the primary header (required only if raw data is passed in)
+        ext_hdr (astropy.io.fits.Header): the image extension header (required only if raw data is passed in)
+     
+    Attrs:
+        value: the getter of the kgain value
+        _kgain (float): the value of kgain
+    """
+    def __init__(self, data_or_filepath, pri_hdr=None, ext_hdr=None):
 
+        # run the image class contructor
+        super().__init__(data_or_filepath, pri_hdr=pri_hdr, ext_hdr=ext_hdr)
+
+        # File format checks
+        if self.data.shape != (1,1):
+            raise ValueError('The KGain calibration data should be just one float value')
+
+        self._kgain = self.data[0,0] 
+        # additional bookkeeping for a calibration file
+        # if this is a new calibration file, we need to bookkeep it in the header
+        # b/c of logic in the super.__init__, we just need to check this to see if it is a new KGain file
+        if ext_hdr is not None:
```

Comment:
If `data_or_filepath` were a file path, then the `Image` class doesn't require pri_hdr or ext_hdr, and then the new headers below are not added to the file, presumably because the user assumes the file represented by the file path already includes these KGain-specific headers.  The error case check in line 597 only applies to the case where a file path is read in.  So if we require a file read in to have the right header at least for 'DATATYPE' (so that not just any file is read in), by the same token we should require an input ext_hdr to already have these KGain-specific headers, in my opinion.  I know this is a thing that is done in all these data classes which inherit from Image in this Python file, but that is my thought for all of them.  Either that, or we simply add these KGain-specific headers in even in the case when a file path is used as the input, which would effectively take away the check that happens at line 597.

Reply:
I am not sure whether I understand what you mean. You are right, line 585 is just a check whether it is a new KGain file with only the data as input. Then the DATATYPE is added to the ext_hdr. It might already be in the ext_hdr input but then it does not harm to overwrite. As you stated the test at line 597 is only needed in case of a file as input. It checks whether there is by chance a wrong calibration file used as input. Maybe @semaphoreP could comment?

Reply:
Okay, after thinking about it, I'm okay with that.  For me, it was really an issue of the philosophy of the check.  If a user can't get away with inputting a file path without a KGain-specific header, then by the same token a user shouldn't be able to enter an array without an ext_hdr that includes a KGain-specific header to minimize the chances of accidentally inputting non-KGain data.  But we aren't being draconian with inputs in general, and I think the typical user of these classes wouldn't mess this up, so I'm okay with this.  And I can understand the convenience of this in practice.